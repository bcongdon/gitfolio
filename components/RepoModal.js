import React from "react";
import { Text, Button, Icon, Card, CardItem, H3, Spinner } from "native-base";
import moment from "moment";
import { Linking, StyleSheet, View } from "react-native";
import nodeEmoji from "node-emoji";
import GitHubColors from "github-colors";
import { GITHUB_DARK_GRAY } from "../constants";
import { VictoryLine } from "victory-native";
import client from "../client";

export default class RepoModal extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      // A list of numbers which represent the average weekly number of commits to the repo
      contributionStats: []
    };
  }

  componentWillMount() {
    this.setState({ starred: this.props.viewerHasStarred });
    this.fetchCommitStats(1000);
  }

  /**
   * Fetches commit statistics from the Github API and updates component state accordingly
   * @param  {number} delay The delay to wait before trying to load commit stats again, if the request fails
   */
  fetchCommitStats(delay) {
    client.getCommitActivity(this.props.owner, this.props.name).then(resp => {
      if (resp.status == 202) {
        setTimeout(() => this.fetchCommitStats(1000 * 1.5), delay);
        return;
      }

      const days = resp.data
        .map(week => week.days.reduce((acc, val) => acc + val), 0)
        .map(weekSum => weekSum / 7);
      this.setState({ contributionStats: days });
    });
  }

  /**
   * Renders an icon tag for a repository
   * @param  {Object} iconOpts Options object to apply to the icon
   * @param  {String} text     Text to display next to the icon
   */
  getRepoIconTag(iconOpts, text) {
    return (
      <Text>
        <Icon {...iconOpts} /> {text}
      </Text>
    );
  }

  /**
   * Renders an icon tag for the provided language
   * @param  {String} language The language
   */
  getLanguageTag(language) {
    const color = GitHubColors.get(language);
    if (!language || !color) {
      return null;
    }

    return this.getRepoIconTag(
      {
        style: { color: color.color, fontSize: 16 },
        type: "MaterialCommunityIcons",
        name: "checkbox-blank-circle"
      },
      language
    );
  }

  /**
   * Renders an icon tag with the provided number of stars
   * @param  {number} stars The number of stars
   */
  getStarsTag(stars) {
    return this.getRepoIconTag(
      {
        style: { fontSize: 16 },
        type: "Octicons",
        name: "star",
        ios: "star"
      },
      stars
    );
  }

  /**
   * Renders an icon tag with the provided number of forks
   * @param  {number} forks The number of forks
   */
  getForksTag(forks) {
    return this.getRepoIconTag(
      {
        style: { fontSize: 16 },
        type: "Octicons",
        name: "repo-forked",
        ios: "repo-forked"
      },
      forks
    );
  }

  /**
   * Renders an icon tag with the the given updated date
   * @param  {Date} pushed_at The "updated_at" date of the repo
   */
  getUpdatedTag(pushed_at) {
    return this.getRepoIconTag(
      {
        style: { fontSize: 16 },
        type: "Octicons",
        name: "calendar",
        ios: "calendar"
      },
      moment(pushed_at).fromNow()
    );
  }

  /**
   * Renders the "star" button for the repo modal
   * @param  {boolean} starred Whether or not the user has starred this repo
   */
  getStarButton(starred) {
    let buttonText = starred ? "Unstar" : "Star";
    return (
      <Button
        style={{ borderColor: GITHUB_DARK_GRAY }}
        small
        iconLeft
        dark
        bordered
        onPress={() => this.toggleRepoStar()}
      >
        <Text style={{ color: GITHUB_DARK_GRAY }}>{buttonText}</Text>
      </Button>
    );
  }

  /**
   * Toggle whether or not the user is starring the current repo
   */
  toggleRepoStar() {
    this.props.onSetStarred(!this.state.starred);
    this.setState({ starred: !this.state.starred });
  }

  /**
   * Renders a button to view the repo in the Github mobile website
   */
  getViewButton() {
    return (
      <Button
        style={{ borderColor: GITHUB_DARK_GRAY, marginRight: 10 }}
        small
        iconLeft
        dark
        bordered
        onPress={() => Linking.openURL(this.props.url)}
      >
        <Text style={{ color: GITHUB_DARK_GRAY }}>{"View"}</Text>
      </Button>
    );
  }

  /**
   * Renders a VictoryLine chart with the contribution statistics
   */
  getCommitsChart() {
    // Return a spinner if no contribution stats have been loaded
    if (this.state.contributionStats.length == 0) {
      return (
        <View style={styles.container}>
          <Spinner />
        </View>
      );
    }

    return (
      <View style={styles.container}>
        <VictoryLine
          height={200}
          interpolation="cardinal"
          style={{
            data: { stroke: "#d6ecb2" }
          }}
          data={this.state.contributionStats.map((val, idx) => {
            return { x: idx, y: val };
          })}
        />
      </View>
    );
  }

  /**
   * Renders a repository as a ListItem
   */
  render() {
    // Conditionally render fields as necessary
    let languageTag = this.props.language
      ? this.getLanguageTag(this.props.language)
      : null;
    let starsTag =
      this.props.stars > 0 ? this.getStarsTag(this.props.stars) : null;
    let forksTag =
      this.props.forks > 0 ? this.getForksTag(this.props.forks) : null;

    return (
      <Card style={{ flex: 0.5 }}>
        <CardItem>
          <Text>
            <H3>{this.props.name}</H3>
            {"\n"}

            <Text note>{nodeEmoji.emojify(this.props.description)}</Text>
            {"\n\n"}

            {languageTag}
            {languageTag ? "  " : null}

            {starsTag}
            {starsTag ? "  " : null}

            {forksTag}
            {forksTag ? "  " : null}

            {this.getUpdatedTag(this.props.pushedAt)}
          </Text>
        </CardItem>
        <CardItem>{this.getCommitsChart()}</CardItem>
        <CardItem>
          {this.getViewButton()}
          {this.getStarButton(this.state.starred)}
        </CardItem>
      </Card>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
    marginTop: 25
  }
});
