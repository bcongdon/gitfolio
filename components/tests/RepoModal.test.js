import { Spinner } from "native-base";
import RepoModal from "../RepoModal";
import React from "react";
import client from "../../client";
import { VictoryLine } from "victory-native";

import renderer from "react-test-renderer";

const createDummyEndpoint = () =>
  jest.genMockFn().mockReturnValue(new Promise(() => {}));

it("renders without crashing", () => {
  client.getCommitActivity = createDummyEndpoint();
  let component = (
    <RepoModal
      name="repo"
      language="Python"
      description="a repo"
      stars={2}
      forks={10}
      viewerHasStarred={false}
    />
  );

  const rendered = renderer.create(component).toJSON();
  expect(rendered).toBeTruthy();
});

it("should snapshot correctly with forks and stars", () => {
  client.getCommitActivity = createDummyEndpoint();
  let component = (
    <RepoModal
      name="repo"
      language="Python"
      description="a repo"
      stars={2}
      forks={10}
      viewerHasStarred={false}
    />
  );

  const rendered = renderer.create(component).toJSON();
  expect(rendered).toMatchSnapshot();
});

it("should snapshot correctly without stars", () => {
  client.getCommitActivity = createDummyEndpoint();
  let component = (
    <RepoModal
      name="repo"
      language="Python"
      description="a repo"
      stars={0}
      forks={10}
      viewerHasStarred={false}
    />
  );

  const rendered = renderer.create(component).toJSON();
  expect(rendered).toMatchSnapshot();
});

it("should snapshot correctly without forks", () => {
  client.getCommitActivity = createDummyEndpoint();
  let component = (
    <RepoModal
      name="repo"
      language="Python"
      description="a repo"
      stars={2}
      forks={10}
      viewerHasStarred={false}
    />
  );

  const rendered = renderer.create(component).toJSON();
  expect(rendered).toMatchSnapshot();
});

it("requests commit activity on mount", () => {
  client.getCommitActivity = createDummyEndpoint();
  let component = (
    <RepoModal
      name="repo"
      language="Python"
      description="a repo"
      owner={{ login: "owner" }}
      stars={2}
      forks={10}
      viewerHasStarred={false}
    />
  );

  renderer.create(component).root;
  expect(client.getCommitActivity).toBeCalled();
  expect(client.getCommitActivity.mock.calls[0][0]).toEqual({ login: "owner" });
  expect(client.getCommitActivity.mock.calls[0][1]).toEqual("repo");
});

it("renders spinner while waiting for commit activity", () => {
  client.getCommitActivity = createDummyEndpoint();
  let component = (
    <RepoModal
      name="repo"
      language="Python"
      description="a repo"
      owner={{ login: "owner" }}
      stars={2}
      forks={10}
      viewerHasStarred={false}
    />
  );

  const rendered = renderer.create(component).root;
  expect(rendered.findByType(Spinner)).toBeTruthy();
});

it("renders VictoryLine after receiving commit activity", () => {
  client.getCommitActivity = createDummyEndpoint();
  let component = (
    <RepoModal
      name="repo"
      language="Python"
      description="a repo"
      owner={{ login: "owner" }}
      stars={2}
      forks={10}
      viewerHasStarred={false}
    />
  );

  const rendered = renderer.create(component).root;
  rendered.instance.setState({ contributionStats: [1, 2, 3] });
  jest.runAllTicks();

  expect(rendered.findByType(VictoryLine)).toBeTruthy();
});
