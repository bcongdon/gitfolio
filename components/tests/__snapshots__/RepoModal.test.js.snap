// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should snapshot correctly with forks and stars 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
        "borderBottomWidth": 0.5,
        "borderColor": "#ccc",
        "borderLeftWidth": 0.5,
        "borderRadius": 2,
        "borderRightWidth": 0.5,
        "borderTopWidth": 0.5,
        "elevation": 3,
        "flex": 1,
        "flexWrap": "nowrap",
        "marginBottom": 5,
        "marginLeft": 2,
        "marginRight": 2,
        "marginTop": 5,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.1,
        "shadowRadius": 1.5,
      },
      Object {
        "flex": 0.5,
      },
    ]
  }
  virtual={undefined}
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "borderRadius": 2,
        "flexDirection": "row",
        "paddingBottom": 10,
        "paddingLeft": 15,
        "paddingRight": 15,
        "paddingTop": 10,
      }
    }
    virtual={undefined}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#000",
          "fontFamily": "System",
          "fontSize": 16,
        }
      }
      uppercase={false}
      virtual={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#000",
            "fontSize": 21,
            "lineHeight": 22,
          }
        }
        virtual={undefined}
      >
        repo
      </Text>
      

      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        note={true}
        style={
          Object {
            "color": "#a7a7a7",
            "fontFamily": "System",
            "fontSize": 14,
          }
        }
        uppercase={false}
        virtual={undefined}
      >
        a repo
      </Text>
      


      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#000",
            "fontFamily": "System",
            "fontSize": 16,
          }
        }
        uppercase={false}
        virtual={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        />
         
        Python
      </Text>
        
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#000",
            "fontFamily": "System",
            "fontSize": 16,
          }
        }
        uppercase={false}
        virtual={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        />
         
        2
      </Text>
        
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#000",
            "fontFamily": "System",
            "fontSize": 16,
          }
        }
        uppercase={false}
        virtual={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        />
         
        10
      </Text>
        
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#000",
            "fontFamily": "System",
            "fontSize": 16,
          }
        }
        uppercase={false}
        virtual={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        />
         
        a few seconds ago
      </Text>
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "borderRadius": 2,
        "flexDirection": "row",
        "paddingBottom": 10,
        "paddingLeft": 15,
        "paddingRight": 15,
        "paddingTop": 10,
      }
    }
    virtual={undefined}
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
          "marginTop": 25,
        }
      }
    >
      <ActivityIndicator
        animating={true}
        color="#45D56E"
        hidesWhenStopped={true}
        size="large"
        style={
          Object {
            "height": 80,
          }
        }
        virtual={undefined}
      />
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "borderRadius": 2,
        "flexDirection": "row",
        "paddingBottom": 10,
        "paddingLeft": 15,
        "paddingRight": 15,
        "paddingTop": 10,
      }
    }
    virtual={undefined}
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "flex-start",
          "backgroundColor": "transparent",
          "borderBottomWidth": 1,
          "borderColor": "#333",
          "borderLeftWidth": 1,
          "borderRadius": 5,
          "borderRightWidth": 1,
          "borderTopWidth": 1,
          "borderWidth": 1,
          "elevation": null,
          "flexDirection": "row",
          "height": 30,
          "justifyContent": "space-between",
          "marginRight": 10,
          "opacity": 1,
          "paddingBottom": 6,
          "paddingTop": 6,
          "shadowColor": null,
          "shadowOffset": null,
          "shadowOpacity": null,
          "shadowRadius": null,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "color": "#000",
              "fontFamily": "System",
              "fontSize": 14,
              "lineHeight": 19,
              "marginLeft": 0,
              "marginRight": 0,
              "paddingLeft": 16,
              "paddingRight": 16,
            },
            Object {
              "color": "#333",
            },
          ]
        }
        uppercase={false}
        virtual={undefined}
      >
        View
      </Text>
    </View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "flex-start",
          "backgroundColor": "transparent",
          "borderBottomWidth": 1,
          "borderColor": "#333",
          "borderLeftWidth": 1,
          "borderRadius": 5,
          "borderRightWidth": 1,
          "borderTopWidth": 1,
          "borderWidth": 1,
          "elevation": null,
          "flexDirection": "row",
          "height": 30,
          "justifyContent": "space-between",
          "opacity": 1,
          "paddingBottom": 6,
          "paddingTop": 6,
          "shadowColor": null,
          "shadowOffset": null,
          "shadowOpacity": null,
          "shadowRadius": null,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "color": "#000",
              "fontFamily": "System",
              "fontSize": 14,
              "lineHeight": 19,
              "marginLeft": 0,
              "marginRight": 0,
              "paddingLeft": 16,
              "paddingRight": 16,
            },
            Object {
              "color": "#333",
            },
          ]
        }
        uppercase={false}
        virtual={undefined}
      >
        Star
      </Text>
    </View>
  </View>
</View>
`;

exports[`should snapshot correctly without forks 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
        "borderBottomWidth": 0.5,
        "borderColor": "#ccc",
        "borderLeftWidth": 0.5,
        "borderRadius": 2,
        "borderRightWidth": 0.5,
        "borderTopWidth": 0.5,
        "elevation": 3,
        "flex": 1,
        "flexWrap": "nowrap",
        "marginBottom": 5,
        "marginLeft": 2,
        "marginRight": 2,
        "marginTop": 5,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.1,
        "shadowRadius": 1.5,
      },
      Object {
        "flex": 0.5,
      },
    ]
  }
  virtual={undefined}
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "borderRadius": 2,
        "flexDirection": "row",
        "paddingBottom": 10,
        "paddingLeft": 15,
        "paddingRight": 15,
        "paddingTop": 10,
      }
    }
    virtual={undefined}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#000",
          "fontFamily": "System",
          "fontSize": 16,
        }
      }
      uppercase={false}
      virtual={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#000",
            "fontSize": 21,
            "lineHeight": 22,
          }
        }
        virtual={undefined}
      >
        repo
      </Text>
      

      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        note={true}
        style={
          Object {
            "color": "#a7a7a7",
            "fontFamily": "System",
            "fontSize": 14,
          }
        }
        uppercase={false}
        virtual={undefined}
      >
        a repo
      </Text>
      


      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#000",
            "fontFamily": "System",
            "fontSize": 16,
          }
        }
        uppercase={false}
        virtual={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        />
         
        Python
      </Text>
        
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#000",
            "fontFamily": "System",
            "fontSize": 16,
          }
        }
        uppercase={false}
        virtual={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        />
         
        2
      </Text>
        
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#000",
            "fontFamily": "System",
            "fontSize": 16,
          }
        }
        uppercase={false}
        virtual={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        />
         
        10
      </Text>
        
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#000",
            "fontFamily": "System",
            "fontSize": 16,
          }
        }
        uppercase={false}
        virtual={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        />
         
        a few seconds ago
      </Text>
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "borderRadius": 2,
        "flexDirection": "row",
        "paddingBottom": 10,
        "paddingLeft": 15,
        "paddingRight": 15,
        "paddingTop": 10,
      }
    }
    virtual={undefined}
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
          "marginTop": 25,
        }
      }
    >
      <ActivityIndicator
        animating={true}
        color="#45D56E"
        hidesWhenStopped={true}
        size="large"
        style={
          Object {
            "height": 80,
          }
        }
        virtual={undefined}
      />
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "borderRadius": 2,
        "flexDirection": "row",
        "paddingBottom": 10,
        "paddingLeft": 15,
        "paddingRight": 15,
        "paddingTop": 10,
      }
    }
    virtual={undefined}
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "flex-start",
          "backgroundColor": "transparent",
          "borderBottomWidth": 1,
          "borderColor": "#333",
          "borderLeftWidth": 1,
          "borderRadius": 5,
          "borderRightWidth": 1,
          "borderTopWidth": 1,
          "borderWidth": 1,
          "elevation": null,
          "flexDirection": "row",
          "height": 30,
          "justifyContent": "space-between",
          "marginRight": 10,
          "opacity": 1,
          "paddingBottom": 6,
          "paddingTop": 6,
          "shadowColor": null,
          "shadowOffset": null,
          "shadowOpacity": null,
          "shadowRadius": null,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "color": "#000",
              "fontFamily": "System",
              "fontSize": 14,
              "lineHeight": 19,
              "marginLeft": 0,
              "marginRight": 0,
              "paddingLeft": 16,
              "paddingRight": 16,
            },
            Object {
              "color": "#333",
            },
          ]
        }
        uppercase={false}
        virtual={undefined}
      >
        View
      </Text>
    </View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "flex-start",
          "backgroundColor": "transparent",
          "borderBottomWidth": 1,
          "borderColor": "#333",
          "borderLeftWidth": 1,
          "borderRadius": 5,
          "borderRightWidth": 1,
          "borderTopWidth": 1,
          "borderWidth": 1,
          "elevation": null,
          "flexDirection": "row",
          "height": 30,
          "justifyContent": "space-between",
          "opacity": 1,
          "paddingBottom": 6,
          "paddingTop": 6,
          "shadowColor": null,
          "shadowOffset": null,
          "shadowOpacity": null,
          "shadowRadius": null,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "color": "#000",
              "fontFamily": "System",
              "fontSize": 14,
              "lineHeight": 19,
              "marginLeft": 0,
              "marginRight": 0,
              "paddingLeft": 16,
              "paddingRight": 16,
            },
            Object {
              "color": "#333",
            },
          ]
        }
        uppercase={false}
        virtual={undefined}
      >
        Star
      </Text>
    </View>
  </View>
</View>
`;

exports[`should snapshot correctly without stars 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
        "borderBottomWidth": 0.5,
        "borderColor": "#ccc",
        "borderLeftWidth": 0.5,
        "borderRadius": 2,
        "borderRightWidth": 0.5,
        "borderTopWidth": 0.5,
        "elevation": 3,
        "flex": 1,
        "flexWrap": "nowrap",
        "marginBottom": 5,
        "marginLeft": 2,
        "marginRight": 2,
        "marginTop": 5,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.1,
        "shadowRadius": 1.5,
      },
      Object {
        "flex": 0.5,
      },
    ]
  }
  virtual={undefined}
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "borderRadius": 2,
        "flexDirection": "row",
        "paddingBottom": 10,
        "paddingLeft": 15,
        "paddingRight": 15,
        "paddingTop": 10,
      }
    }
    virtual={undefined}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#000",
          "fontFamily": "System",
          "fontSize": 16,
        }
      }
      uppercase={false}
      virtual={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#000",
            "fontSize": 21,
            "lineHeight": 22,
          }
        }
        virtual={undefined}
      >
        repo
      </Text>
      

      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        note={true}
        style={
          Object {
            "color": "#a7a7a7",
            "fontFamily": "System",
            "fontSize": 14,
          }
        }
        uppercase={false}
        virtual={undefined}
      >
        a repo
      </Text>
      


      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#000",
            "fontFamily": "System",
            "fontSize": 16,
          }
        }
        uppercase={false}
        virtual={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        />
         
        Python
      </Text>
        
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#000",
            "fontFamily": "System",
            "fontSize": 16,
          }
        }
        uppercase={false}
        virtual={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        />
         
        10
      </Text>
        
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#000",
            "fontFamily": "System",
            "fontSize": 16,
          }
        }
        uppercase={false}
        virtual={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        />
         
        a few seconds ago
      </Text>
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "borderRadius": 2,
        "flexDirection": "row",
        "paddingBottom": 10,
        "paddingLeft": 15,
        "paddingRight": 15,
        "paddingTop": 10,
      }
    }
    virtual={undefined}
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
          "marginTop": 25,
        }
      }
    >
      <ActivityIndicator
        animating={true}
        color="#45D56E"
        hidesWhenStopped={true}
        size="large"
        style={
          Object {
            "height": 80,
          }
        }
        virtual={undefined}
      />
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "borderRadius": 2,
        "flexDirection": "row",
        "paddingBottom": 10,
        "paddingLeft": 15,
        "paddingRight": 15,
        "paddingTop": 10,
      }
    }
    virtual={undefined}
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "flex-start",
          "backgroundColor": "transparent",
          "borderBottomWidth": 1,
          "borderColor": "#333",
          "borderLeftWidth": 1,
          "borderRadius": 5,
          "borderRightWidth": 1,
          "borderTopWidth": 1,
          "borderWidth": 1,
          "elevation": null,
          "flexDirection": "row",
          "height": 30,
          "justifyContent": "space-between",
          "marginRight": 10,
          "opacity": 1,
          "paddingBottom": 6,
          "paddingTop": 6,
          "shadowColor": null,
          "shadowOffset": null,
          "shadowOpacity": null,
          "shadowRadius": null,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "color": "#000",
              "fontFamily": "System",
              "fontSize": 14,
              "lineHeight": 19,
              "marginLeft": 0,
              "marginRight": 0,
              "paddingLeft": 16,
              "paddingRight": 16,
            },
            Object {
              "color": "#333",
            },
          ]
        }
        uppercase={false}
        virtual={undefined}
      >
        View
      </Text>
    </View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "flex-start",
          "backgroundColor": "transparent",
          "borderBottomWidth": 1,
          "borderColor": "#333",
          "borderLeftWidth": 1,
          "borderRadius": 5,
          "borderRightWidth": 1,
          "borderTopWidth": 1,
          "borderWidth": 1,
          "elevation": null,
          "flexDirection": "row",
          "height": 30,
          "justifyContent": "space-between",
          "opacity": 1,
          "paddingBottom": 6,
          "paddingTop": 6,
          "shadowColor": null,
          "shadowOffset": null,
          "shadowOpacity": null,
          "shadowRadius": null,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "color": "#000",
              "fontFamily": "System",
              "fontSize": 14,
              "lineHeight": 19,
              "marginLeft": 0,
              "marginRight": 0,
              "paddingLeft": 16,
              "paddingRight": 16,
            },
            Object {
              "color": "#333",
            },
          ]
        }
        uppercase={false}
        virtual={undefined}
      >
        Star
      </Text>
    </View>
  </View>
</View>
`;
